classpath/java/net/Socket.java

```diff
@@ -93,7 +93,9 @@ public int read(byte[] buffer) throws IOException {
 				size = recv(sock, buffer, index, Math.min(fullSize, Socket.BUFFER_SIZE));
 				fullSize -= size;
 				index += size;
-			} while (fullSize != 0 && size != 0);
+			} while (fullSize != 0 && size > 0);
+			if(size < 0) throw new IOException("Error whilereading stream");
+			if(index == 0 && size == 0) return -1;
```

Comment:
There's the edge case here where `buffer.length == 0`.  Intuitively (and according to the spec: http://docs.oracle.com/javase/7/docs/api/java/io/InputStream.html#read(byte[]) ) - but if I'm reading this right, this code will return -1 in that case.